mutable_tree.go

```diff
@@ -442,80 +458,123 @@ func (tree *MutableTree) GetVersioned(key []byte, version int64) (
 // If version is snapshot version, persist version to disk as well
 func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
 	version := tree.version + 1
+	vm := &VersionMetadata{
+		Version:  version,
+		Snapshot: tree.ndb.opts.KeepEvery != 0 && version%tree.ndb.opts.KeepEvery == 0,
+	}
 
 	if tree.versions[version] {
-		//version already exists, throw an error if attempting to overwrite
-		// Same hash means idempotent.  Return success.
+		// If the version already exists, return an error as we're attempting to overwrite.
+		// However, the same hash means idempotent (i.e. no-op).
 		existingHash, err := tree.ndb.getRoot(version)
 		if err != nil {
 			return nil, version, err
 		}
+
 		var newHash = tree.WorkingHash()
+
 		if bytes.Equal(existingHash, newHash) {
 			tree.version = version
 			tree.ImmutableTree = tree.ImmutableTree.clone()
 			tree.lastSaved = tree.ImmutableTree.clone()
 			tree.orphans = map[string]int64{}
 			return existingHash, version, nil
 		}
-		return nil, version, fmt.Errorf("version %d was already saved to different hash %X (existing hash %X)",
-			version, newHash, existingHash)
+
+		return nil, version, fmt.Errorf("version %d was already saved to different hash %X (existing hash %X)", version, newHash, existingHash)
 	}
+
 	tree.versions[version] = true
 
 	if tree.root == nil {
 		// There can still be orphans, for example if the root is the node being
 		// removed.
 		debug("SAVE EMPTY TREE %v\n", version)
-		tree.ndb.SaveOrphans(version, tree.orphans)
-		err := tree.ndb.SaveEmptyRoot(version)
-		if err != nil {
+		if err := tree.ndb.SaveOrphans(version, tree.orphans); err != nil {
+			panic(err)
+		}
+
+		if err := tree.ndb.SaveEmptyRoot(version); err != nil {
 			panic(err)
 		}
 	} else {
 		debug("SAVE TREE %v\n", version)
 		// Save the current tree.
-		tree.ndb.SaveTree(tree.root, version)
-		tree.ndb.SaveOrphans(version, tree.orphans)
-		err := tree.ndb.SaveRoot(tree.root, version)
-		if err != nil {
+
+		if _, err := tree.ndb.SaveTree(tree.root, version); err != nil {
+			panic(err)
+		}
+
+		if err := tree.ndb.SaveOrphans(version, tree.orphans); err != nil {
+			panic(err)
+		}
+
+		if err := tree.ndb.SaveRoot(tree.root, version); err != nil {
 			panic(err)
 		}
-	}
-	err := tree.ndb.Commit()
```

Comment:
Let's keep this commit, removing it may cause higher memory usage since the in-memory batches now have to hold both the queued changes and the pruning changes.

Reply:
The commit is kept. It happens in `MutableTree#PruneRecentVersion`.

Reply:
Got it, I misread the diff.